Change WebPage::Reload to use OptionSet<>

abigailfox opened the pull request:
#### 816f1ad80041069759cf50ad770f72aabbb8d033
<pre>
Change WebPage::Reload to use OptionSet&lt;&gt;
<a href="https://bugs.webkit.org/show_bug.cgi?id=244313">https://bugs.webkit.org/show_bug.cgi?id=244313</a>
rdar://63370308

Reviewed by NOBODY (OOPS!).

        modified:   Source/WebCore/loader/FrameLoaderTypes.h
        modified:   Source/WebKit/Scripts/webkit/messages.py
        modified:   Source/WebKit/UIProcess/WebPageProxy.cpp
        modified:   Source/WebKit/WebProcess/WebPage/WebPage.cpp
        modified:   Source/WebKit/WebProcess/WebPage/WebPage.h
        modified:   Source/WebKit/WebProcess/WebPage/WebPage.messages.in
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e9fc969f3c5b5971f5f368edfdf1bd0dd1603e2d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87204 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31291 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18032 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/96237 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149780 "failed Failed to checkout and rebase branch from PR 3666") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29655 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91216 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92820 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23958 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/74009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/79320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78934 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/79179 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/36/builds/79320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27380 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/13016 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27327 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/14031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29012 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/74009 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28951 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/33308 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/loader/FrameLoaderTypes.h:
```diff
@@ -271,4 +271,13 @@ template<> struct EnumTraits<WebCore::WebGLLoadPolicy> {
     >;
 };
 
+template<> struct EnumTraits<WebCore::ReloadOption> {
```
We could move this up after `EnumTraits<WebCore::PolicyAction>` to pseudo-alphabetize the `EnumTraits`, but not strictly necessary.

ddkilzer reviewed (approved):
r=me, but consider moving the `EnumTraits<WebCore::PolicyAction>` definition.

abigailfox committed:
Change WebPage::Reload to use OptionSet<>
https://bugs.webkit.org/show_bug.cgi?id=244313
rdar://63370308

Reviewed by NOBODY (OOPS!).

        modified:   Source/WebCore/loader/FrameLoaderTypes.h
        modified:   Source/WebKit/Scripts/webkit/messages.py
        modified:   Source/WebKit/UIProcess/WebPageProxy.cpp
        modified:   Source/WebKit/WebProcess/WebPage/WebPage.cpp
        modified:   Source/WebKit/WebProcess/WebPage/WebPage.h
        modified:   Source/WebKit/WebProcess/WebPage/WebPage.messages.in

karlrackler committed:
[Gardening]: [ macOS wk1 ] imported/w3c/web-platform-tests/intersection-observer/target-in-different-window.html is a flaky failure
https://bugs.webkit.org/show_bug.cgi?id=244372
<rdar://99163929>

Unreviewed test gardening.

* LayoutTests/platform/mac-wk1/TestExpectations:

Canonical link: https://commits.webkit.org/253800@main

alanbaradlay committed:
[IFC][Integration] Introduce InlineContentPainter
https://bugs.webkit.org/show_bug.cgi?id=244339

Reviewed by Antti Koivisto.

Move painting logic from LineLayout to InlineContentPainter.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/layout/integration/inline/LayoutIntegrationInlineContentPainter.cpp: Added.
(WebCore::LayoutIntegration::InlineContentPainter::InlineContentPainter):
(WebCore::LayoutIntegration::LayerPaintScope::LayerPaintScope):
(WebCore::LayoutIntegration::LayerPaintScope::includes):
(WebCore::LayoutIntegration::InlineContentPainter::paint):
(WebCore::LayoutIntegration::InlineContentPainter::flippedContentOffsetIfNeeded const):
* Source/WebCore/layout/integration/inline/LayoutIntegrationInlineContentPainter.h: Added.
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.cpp:
(WebCore::LayoutIntegration::LineLayout::paint):
(WebCore::LayoutIntegration::LayerPaintScope::LayerPaintScope): Deleted.
(WebCore::LayoutIntegration::LayerPaintScope::includes): Deleted.
(): Deleted.
* Source/WebCore/layout/integration/inline/LayoutIntegrationLineLayout.h:

Canonical link: https://commits.webkit.org/253801@main

megangardner committed:
Fragment Directive not removed from URL.
https://bugs.webkit.org/show_bug.cgi?id=244318

Reviewed by Ryosuke Niwa and Alex Christensen.

We were not stripping the fragment directive from the URL when setting the URL on the document.
This is necessary to be spec compliant as well as we would like to maintain user privacy and not
allow Text Fragment to become a tracking vector. Adding the WPT that now passes since all the
URL is stripped out, which was not the main thing that was being tested, but it now completely
passes where other tests still have failures I'm working through.

* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-same-doc-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/scroll-to-text-fragment/scroll-to-text-fragment-same-doc.html: Added.
* Source/WTF/wtf/URL.cpp:
(WTF::URL::consumefragmentDirective):
* Source/WTF/wtf/URL.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::setURL):
* Source/WebCore/dom/Document.h:
(WebCore::Document::fragmentDirective const):
* Source/WebCore/dom/FragmentDirectiveParser.cpp:
(WebCore::FragmentDirectiveParser::FragmentDirectiveParser):
* Source/WebCore/dom/FragmentDirectiveParser.h:
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::scrollToFragment):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::getTextFragmentMatch):

Canonical link: https://commits.webkit.org/253802@main

mwyrzykowski committed:
Computed regions appear incorrect in fullscreen
https://bugs.webkit.org/show_bug.cgi?id=244056
<rdar://97562328>

Reviewed by Tim Nguyen.

When in fullscreen, only allow elements under the fullscreen
element to be non-inert. That effectively disables them
from interaction during fullscreen.

* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):

Canonical link: https://commits.webkit.org/253803@main

Commit:
Let Xcode have its way with project files
<https://bugs.webkit.org/show_bug.cgi?id=244373>

Unreviewed Xcode gardening.

* Source/ThirdParty/libwebrtc/libwebrtc.xcodeproj/project.pbxproj:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
- Let Xcode resort items that were merged out-of-order.

Canonical link: https://commits.webkit.org/253804@main

Constellation committed:
Use C++20 destroying operator delete for NodeRareData
https://bugs.webkit.org/show_bug.cgi?id=244325

Reviewed by Ryosuke Niwa.

C++20 introduces destroying operator delete which can introduce a hook-point before calling a destructor.
This allows us to dispatch destructors by reading some data from the target without using `virtual` destructor.
This patch applies it to NodeRareData so that we no longer need to have custom deleter for NodeRareData.

* Source/WebCore/dom/Node.cpp:
(WebCore::NodeRareData::operator delete):
(WebCore::Node::materializeRareData):
(WebCore::Node::NodeRareDataDeleter::operator() const): Deleted.
* Source/WebCore/dom/Node.h:
* Source/WebCore/dom/NodeRareData.h:

Canonical link: https://commits.webkit.org/253805@main

cdumez committed:
TEST ADD (250224@main): [ macOS wk1 Debug ] storage/domstorage/sessionstorage/window-open-remove-item.html is a consistent crash
https://bugs.webkit.org/show_bug.cgi?id=244338
<rdar://99143234>

Reviewed by Sihui Liu.

Make sure we don't call window.open() before the load event has fired since DRT
doesn't like that.

* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/storage/domstorage/sessionstorage/window-open-remove-item.html:

Canonical link: https://commits.webkit.org/253806@main

zdobersek committed:
[CoordinatedGraphics] Cache and reuse image-based backing stores
https://bugs.webkit.org/show_bug.cgi?id=242394

Reviewed by Miguel Gomez.

Implement caching of image-based backing stores in the CoordinatedGraphics
subsystem. This allows for such backing stores to be shared between layers that
reuse the same image resource as its content.

The CompositingCoordinator object, in control of a given layer tree, now also
holds a cache of image-based backing store objects that can be shared between
different layers using the same image resource as its content. This cache is
then mirrored in CoordinatedGraphicsScene, but there the CoordinatedBackingStore
objects are cached instead, which are created with the desired image resource
upon entry into the cache. This way different layer objects can end up sharing
on-GPU memory for display of the desired image content, and the CPU-bound image
memory consumption can also be reduced.

* Source/WebCore/platform/TextureMapper.cmake:
* Source/WebCore/platform/graphics/nicosia/texmap/NicosiaImageBackingStore.cpp: Copied from Source/WebCore/platform/graphics/nicosia/texmap/NicosiaImageBackingTextureMapperImpl.cpp.
* Source/WebCore/platform/graphics/nicosia/texmap/NicosiaImageBackingStore.h: Copied from Source/WebCore/platform/graphics/nicosia/texmap/NicosiaImageBackingTextureMapperImpl.cpp.
(Nicosia::ImageBackingStore::backingStoreState):
(Nicosia::ImageBackingStore::CompositionState::CompositionState):
(Nicosia::ImageBackingStore::compositionState):
* Source/WebCore/platform/graphics/nicosia/texmap/NicosiaImageBackingTextureMapperImpl.cpp:
(Nicosia::ImageBackingTextureMapperImpl::flushUpdate):
* Source/WebCore/platform/graphics/nicosia/texmap/NicosiaImageBackingTextureMapperImpl.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.cpp:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStore.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.cpp:
(WebCore::CoordinatedGraphicsLayer::flushCompositingStateForThisLayerOnly):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedGraphicsLayer.h:
* Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::updateImageBacking):
(WebKit::removeLayer):
(WebKit::CoordinatedGraphicsScene::updateSceneState):
(WebKit::CoordinatedGraphicsScene::purgeGLResources):
* Source/WebKit/Shared/CoordinatedGraphics/CoordinatedGraphicsScene.h:
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.cpp:
(WebKit::CompositingCoordinator::flushPendingLayerChanges):
(WebKit::CompositingCoordinator::imageBackingStore):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CompositingCoordinator.h:

Canonical link: https://commits.webkit.org/253807@main

kkinnunen-apple committed:
Cocoa code in Source/WebKit/Platform is not available in C++ unit tests
https://bugs.webkit.org/show_bug.cgi?id=243985
rdar://problem/98715202

Reviewed by Elliott Williams.

Link Source/WebKit/Platform to libWebKitPlatform.a static library.
The C++ unit tests can then directly link to that library.

Before,
  The code in Source/WebKit/Platform/* was compiled as part
  of UnifiedSource??.cpp files.
  The objects were linked to WebKit.framework.

After,
  The code in Source/WebKit/Platform/* is compiled as part
  of PlatformUnifiedSource??.cpp files.
  The objects are linked to libWebKitPlatform.a static library.
  The static library is linked to WebKit.framework.
  The rest of the Source/WebKit code is compiled as before.

Leaves few source files in Source/WebKit/Platform/* to be compiled
directly to WebKit.framework, as those are really not part of
Source/WebKit/Platform. They are using code from other
parts of Source/WebKit.

Later on, TestIPC executable will link to libWebKitPlatform.a and include
headers directly from the Source/WebKit/Platform/* implementation path.
This way TestIPC is able to test all the IPC code.

This satisfies following requirements:
- Does not add new testing related symbols to WebKit.framework. WebKit.framework
  is required to export mostly the WK API and nothing more.
- Does not add new testing related headers to WebKit.framework public
  or private headers. WebKit.framework headers in <WebKit/> are intended
  to be the public WK headers only.
- Is able to test private, non-exported symbols of Source/WebKit/Platform/*
- The IPC code depends on rest of the WebKit/Platform code, and as such
  requires libWebKitPlatform as opposed to hypothetical libWebKitIPC.

Uses a new unified source list in Source/WebKit/Platform/Sources*.txt to
define the WebKitPlatform "component". This keeps all platforms in sync,
regardless whether they build the libWebKitPlatform.a separately or not.

Fixes few unified build include omissions.

Moves the "Copy Profiling Data" build phase from WebKit.framework
target into "Derived Sources" target. The profiling data is
part of the DerivedSources directory and is used by WebKit/Platform
compile.

* Source/WTF/Scripts/generate-unified-source-bundles.rb:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/Platform/IPC/StreamConnectionBuffer.cpp:
* Source/WebKit/Platform/IPC/StreamConnectionBuffer.h:
* Source/WebKit/Platform/Sources.txt: Added.
* Source/WebKit/Platform/SourcesCocoa.txt: Added.
* Source/WebKit/Platform/cocoa/SharedMemoryCocoa.cpp:
(WebKit::SharedMemory::Handle::setOwnershipOfMemory const):
* Source/WebKit/PlatformMac.cmake:
* Source/WebKit/Scripts/generate-unified-sources.sh:
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UnifiedSources-output.xcfilelist:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/Configurations/WebKitPlatform.xcconfig: Added.

Canonical link: https://commits.webkit.org/253808@main

smfr committed:
Fixed position elements should layout relative to transformed container
https://bugs.webkit.org/show_bug.cgi?id=118635

Reviewed by Alan Bujtas.

Layout code already handled `transform` and `will-change: transform` triggering containing block for
`position: fixed`. However, compositing code would still treat the elements as fixed, registering
them with the scrolling tree, so that scrolling would reposition them.

Fix this by having `RenderLayerCompositor::isViewportConstrainedFixedOrStickyLayer()` consult the
`behavesAsFixed()` flag on RenderLayer (rather than the position style).

However, we also need to make `behavesAsFixed()` correct under a `will-change` ancestor, so have
`RenderLayer::recursiveUpdateLayerPositions()` consult `canContainFixedPositionObjects()` and
maintain a new `m_hasFixedContainingBlockAncestor` flag on RenderLayer (since the
`m_hasTransformedAncestor`) one is used for other purposes.

* LayoutTests/scrollingcoordinator/scrolling-tree/fixed-position-within-transformed-expected.txt: Added.
* LayoutTests/platform/ios-wk2/scrollingcoordinator/scrolling-tree/fixed-position-within-transformed-expected.txt: Added.
* LayoutTests/scrollingcoordinator/scrolling-tree/fixed-position-within-transformed.html: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::flagsForUpdateLayerPositions):
(WebCore::RenderLayer::recursiveUpdateLayerPositions):
* Source/WebCore/rendering/RenderLayer.h:
(WebCore::RenderLayer::hasFixedContainingBlockAncestor const):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::isViewportConstrainedFixedOrStickyLayer const):

Canonical link: https://commits.webkit.org/253809@main

zdobersek committed:
[WPE] Temporarily mark media-related tests as slow
https://bugs.webkit.org/show_bug.cgi?id=244379

Unreviewed gardening, media-related tests are marked as slow on WPE while the
issue of modern media controls causing rendering delays due to backdrop filters
is being resolved.

* LayoutTests/platform/wpe/TestExpectations:

Canonical link: https://commits.webkit.org/253810@main

alanbaradlay committed:
[LFC][IFC] shouldWrapUnbreakableContentToNextLine should use the parent style
https://bugs.webkit.org/show_bug.cgi?id=244370

Reviewed by Antti Koivisto.

This change turn Result::Action::RevertToLastWrapOpportunity into a Result::Action::Wrap. While it has no layout implication (last wrap opportunity is right at the end of the committed content), Action::Wrap is more performant since RevertToLastWrapOpportunity forces us to rebuild the line from scratch.

* Source/WebCore/layout/formattingContexts/inline/InlineContentBreaker.cpp:
(WebCore::Layout::InlineContentBreaker::processOverflowingContent const):

Canonical link: https://commits.webkit.org/253811@main

youennf committed:
Log truncate file failure in createMappedFileData
https://bugs.webkit.org/show_bug.cgi?id=244334
rdar://problem/99140113

Reviewed by David Kilzer.

Do a small refactoring to reduce explicit closure of file.
Use RELEASE_LOG_FAULT in case truncate fails.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::createMappedFileData):

Canonical link: https://commits.webkit.org/253812@main

anttijk committed:
[LFC][Integration] Allow 'clear' on all elements
https://bugs.webkit.org/show_bug.cgi?id=244383
<rdar://99176111>

Reviewed by Alan Bujtas.

* Source/WebCore/layout/formattingContexts/inline/InlineFormattingGeometry.cpp:
(WebCore::Layout::InlineFormattingGeometry::logicalTopForNextLine const):
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:

Remove the test.

(WebCore::LayoutIntegration::canUseForLineLayoutWithReason):
* Source/WebCore/layout/layouttree/LayoutBox.cpp:
(WebCore::Layout::Box::hasFloatClear const):

'clear' property only applies to block level boxes (per spec) and linebreak boxes (in practice).

Canonical link: https://commits.webkit.org/253813@main

alanbaradlay committed:
[LFC][IFC] LineBuilder::rebuildLine should return the number of inline boxes attached to the current line
https://bugs.webkit.org/show_bug.cgi?id=244366

Reviewed by Antti Koivisto.

Ignore the float boxes when counting the number of _inline_ boxes placed on the current line.

* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
(WebCore::Layout::LineBuilder::handleInlineContent):
(WebCore::Layout::LineBuilder::rebuildLineWithInlineContent):
(WebCore::Layout::LineBuilder::rebuildLineForTrailingSoftHyphen):
(WebCore::Layout::LineBuilder::rebuildLine): Deleted.
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.h:

Canonical link: https://commits.webkit.org/253814@main

anttijk committed:
[LFC][Integration] Remove LFC preferred width computation define
https://bugs.webkit.org/show_bug.cgi?id=244387
<rdar://99180654>

Reviewed by Alan Bujtas.

* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::invalidateLineLayoutPath):
(WebCore::RenderBlockFlow::tryComputePreferredWidthsUsingModernPath):

Canonical link: https://commits.webkit.org/253815@main

nikolaszimmermann committed:
[LBSE] Add support for 'foreignObject' elements
https://bugs.webkit.org/show_bug.cgi?id=243498

Reviewed by Rob Buis.

Rename RenderSVGForeignObject -> LegacyRenderSVGForeignObject.

Re-introduce RenderSVGForeignObject for LBSE. This marks the finish
of the renderer conversion: all non-resource related renderers have both a legacy
and LBSE Renderer class now: <foreignObject> was the last missing element in LBSE.

Covered by existing svg/foreignObject tests, which all pass now in LBSE (except
three that depends on clipping / masking / filters).

* LayoutTests/platform/mac-monterey-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/background-render-phase-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/body-background-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/clip-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/disallowed-svg-nodes-as-direct-children-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-display-none-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-display-none-with-relative-pos-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-display-changes-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-display-none-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-display-none-with-relative-pos-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-of-parent-display-none-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/fO-parent-of-parent-display-none-with-relative-pos-content-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/filter-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/mask-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/multiple-foreign-objects-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/no-crash-with-svg-content-in-html-document-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/no-crash-with-svg-content-in-html-document-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/svg-document-as-direct-child-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/svg-document-in-html-document-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/svg-document-in-html-document-expected.txt: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/text-tref-02-b-expected.png: Added.
* LayoutTests/platform/mac-monterey-wk2-lbse-text/svg/foreignObject/text-tref-02-b-expected.txt: Added.
* LayoutTests/platform/mac/svg/foreignObject/fO-parent-display-changes-expected.txt:
* LayoutTests/svg/foreignObject/fO-parent-display-changes.svg:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AccessibilitySVGElement.cpp:
(WebCore::AccessibilitySVGElement::determineAccessibilityRole):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::didAttachChild):
(WebCore::RenderElement::layerCreationAllowedForSubtree const):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::canContainFixedPositionObjects const):
(WebCore::RenderElement::canContainAbsolutelyPositionedObjects const):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::updateAncestorDependentState):
(WebCore::RenderLayer::paintForegroundForFragments):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::shouldPaintSVGRenderer const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isLegacySVGForeignObject const):
(WebCore::RenderObject::isSVGForeignObjectOrLegacySVGForeignObject const):
* Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.cpp: Copied from Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp.
(WebCore::LegacyRenderSVGForeignObject::LegacyRenderSVGForeignObject):
(WebCore::LegacyRenderSVGForeignObject::foreignObjectElement const):
(WebCore::LegacyRenderSVGForeignObject::paint):
(WebCore::LegacyRenderSVGForeignObject::localToParentTransform const):
(WebCore::LegacyRenderSVGForeignObject::updateLogicalWidth):
(WebCore::LegacyRenderSVGForeignObject::computeLogicalHeight const):
(WebCore::LegacyRenderSVGForeignObject::layout):
(WebCore::LegacyRenderSVGForeignObject::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/LegacyRenderSVGForeignObject.h: Copied from Source/WebCore/rendering/svg/RenderSVGForeignObject.h.
* Source/WebCore/rendering/svg/LegacyRenderSVGModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGBlock.cpp:
(WebCore::RenderSVGBlock::updateFromStyle):
(WebCore::RenderSVGBlock::mapLocalToContainer const):
(WebCore::RenderSVGBlock::pushMappingToContainer const):
* Source/WebCore/rendering/svg/RenderSVGBlock.h:
* Source/WebCore/rendering/svg/RenderSVGForeignObject.cpp:
(WebCore::RenderSVGForeignObject::paint):
(WebCore::RenderSVGForeignObject::updateLogicalWidth):
(WebCore::RenderSVGForeignObject::computeLogicalHeight const):
(WebCore::RenderSVGForeignObject::layout):
(WebCore::RenderSVGForeignObject::overflowClipRect const):
(WebCore::RenderSVGForeignObject::updateFromStyle):
(WebCore::RenderSVGForeignObject::applyTransform const):
(WebCore::RenderSVGForeignObject::localToParentTransform const): Deleted.
(WebCore::RenderSVGForeignObject::nodeAtFloatPoint): Deleted.
* Source/WebCore/rendering/svg/RenderSVGForeignObject.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
* Source/WebCore/rendering/svg/SVGContainerLayout.cpp:
* Source/WebCore/svg/SVGElement.cpp:
(WebCore::createSVGLayerAwareElementSet):
* Source/WebCore/svg/SVGForeignObjectElement.cpp:
(WebCore::SVGForeignObjectElement::createElementRenderer):
(WebCore::SVGForeignObjectElement::rendererIsNeeded):

Canonical link: https://commits.webkit.org/253816@main

whsieh committed:
[macOS] Refactor some code for reading file paths in PlatformPasteboard
https://bugs.webkit.org/show_bug.cgi?id=244360
rdar://98414238

Reviewed by Devin Rousso.

Add some checks in `getPathnamesForType`, to verify that the incoming pasteboard type is one of the
expected file-path-related pasteboard types.

* Source/WebCore/platform/mac/PlatformPasteboardMac.mm:
(WebCore::isFilePasteboardType):
(WebCore::canWritePasteboardType):
(WebCore::canWriteAllPasteboardTypes):
(WebCore::PlatformPasteboard::getPathnamesForType const):
(WebCore::PlatformPasteboard::addTypes):
(WebCore::PlatformPasteboard::setTypes):

Canonical link: https://commits.webkit.org/253817@main

karlrackler committed:
[Gardening]: REGRESSION (252564@main): [ macOS wk1 Debug ] editing/selection/cleared-by-relayout.html is a consistent crash
https://bugs.webkit.org/show_bug.cgi?id=244390
<rdar://99188200>

Unreviewed test gardening.

* LayoutTests/platform/mac-wk1/TestExpectations:

Canonical link: https://commits.webkit.org/253818@main

litherum committed:
[Cocoa] Small caps should not be synthesized if _any_ character in the font supports small caps
https://bugs.webkit.org/show_bug.cgi?id=244236
<rdar://problem/99004636>

Reviewed by Tim Nguyen.

Previously, we were synthesizing small caps if any character in the run didn't support small caps.
However, the [spec](https://drafts.csswg.org/css-fonts-4/#font-variant-caps-prop) says:

> The availability of these glyphs is based on whether a given feature is defined or not in the feature
> list of the font. User agents ... should explicitly not decide this on a per-character basis.

This patch implements that. The new policy is that if a font supports _any_ small caps characters, then
we don't synthesize any runs with that font.

Tests: imported/w3c/web-platform-tests/css/css-fonts/font-variant-01.html
       imported/w3c/web-platform-tests/css/css-fonts/font-variant-caps-03.html
       imported/w3c/web-platform-tests/css/css-fonts/font-variant-caps-05.html
       imported/w3c/web-platform-tests/css/css-fonts/font-variant-caps.html

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::shouldSynthesize):
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::variantCapsSupportedForSynthesis const):
(WebCore::Font::variantCapsSupportsCharacterForSynthesis const): Deleted.
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::variantCapsSupportedForSynthesis const):
(WebCore::Font::supportsSmallCaps const):
(WebCore::Font::supportsAllSmallCaps const):
(WebCore::Font::supportsPetiteCaps const):
(WebCore::Font::supportsAllPetiteCaps const):
(WebCore::Font::variantCapsSupportsCharacterForSynthesis const): Deleted.
(WebCore::Font::glyphsSupportedBySmallCaps const): Deleted.
(WebCore::Font::glyphsSupportedByAllSmallCaps const): Deleted.
(WebCore::Font::glyphsSupportedByPetiteCaps const): Deleted.
(WebCore::Font::glyphsSupportedByAllPetiteCaps const): Deleted.
* Source/WebCore/platform/graphics/freetype/SimpleFontDataFreeType.cpp:
* LayoutTests/TestExpectations:
* LayoutTests/css3/font-variant-synthesis-jdaggett.html: Replaced by imported/w3c/web-platform-tests/css/css-fonts/font-variant-caps.html
* LayoutTests/css3/font-variant-synthesis-jdaggett-expected.html:
* LayoutTests/platform/gtk/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* LayoutTests/platform/gtk/TestExpectations:

Canonical link: https://commits.webkit.org/253819@main

alanbaradlay committed:
[IFC][Integration] Move display box painting to a dedicated function
https://bugs.webkit.org/show_bug.cgi?id=244344

Reviewed by Antti Koivisto.

This is in preparation for adding ellipsis content painting.

Canonical link: https://commits.webkit.org/253820@main

rwlbuis committed:
[content-visibility] Parse and add experimental flag for content-visibility
https://bugs.webkit.org/show_bug.cgi?id=236371

Reviewed by Darin Adler.

Remove some code committed in PR #3158 that is not needed after r252720 landed.

* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::operator==):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::CSSPropertyParser::parseSingleValue):

Canonical link: https://commits.webkit.org/253821@main

alanbaradlay committed:
[LFC][IFC] Fix shouldApplyTrailingWhiteSpaceFollowedByBRQuirk
https://bugs.webkit.org/show_bug.cgi?id=244375

Reviewed by Antti Koivisto.

shouldApplyTrailingWhiteSpaceFollowedByBRQuirk only applies when the content is not overflowing the line.
When we don't trim this whitespace content in quirk mode:
1. the root inline box gets marked as "has content"
2. "has content" inline boxes affect the line box height
3. line box gets stretched by the root inline box (in this case by the descent).
-> block box becomes 4px taller.

* LayoutTests/fast/text/legacy-trailing-br-quirk-with-overflowing-content-expected.html: Added.
* LayoutTests/fast/text/legacy-trailing-br-quirk-with-overflowing-content.html: Added.
* Source/WebCore/layout/formattingContexts/inline/InlineLineBuilder.cpp:
(WebCore::Layout::LineBuilder::close):
* Source/WebCore/layout/integration/LayoutIntegrationCoverage.cpp:

Canonical link: https://commits.webkit.org/253822@main

vitorroriz committed:
REGRESSION (253685@main): [ Monterey wk2 ] fast/images/text-recognition/mac/cursor-types-for-recognized-text.html is a consistent failure
https://bugs.webkit.org/show_bug.cgi?id=244272
<rdar://99051044>

Reviewed by Wenson Hsieh.

Test was failing because (253685@main) sets cursor:pointer for <a> with links,
since this behavior can't be obtained from cursor:auto according to specification.

Our live-text feature creates a overlay (a <div>) on the shadow DOM tree that depends on
cursor:auto behavior to render a IBeam cursor while hovering the text on top
of the image on the overlay. In this test, we had a <img> inside a <a>. After (253685@main)
the div overlay was inheriting cursor:pointer from the <a> and therefore, we would not
render a IBeam cursor for it.

Solution: We are now styling the div#image-overlay in imageOverlay.css with cursor:auto.
In that way, user can still style the wrapper image, and we preserve cursor:auto by default
for div#image-overlay.

* LayoutTests/fast/images/text-recognition/mac/cursor-types-for-recognized-text-expected.txt:
* LayoutTests/fast/images/text-recognition/mac/cursor-types-for-recognized-text.html:
testImage now tests the cursor hovering on the image area out of the text area.
testOverlayDivFromImage tests the cursor hovering the overlay div on the text area.

We have now 6 tests:
1. cursor on the image inside a link, out of text area, should be "Hand".
2. cursor on the image, with custom cursor, out of text area, should custom cursor.
3. cursor on the image, with user-select-none, inside a link, should be "Hand".
4. cursor on text area of overlay from image [1] should be "IBeam" (hinting selection).
5. cursor on text area of overlay from image [2] should be "IBeam" (hinting selection).
6. cursor on text area of overlay from image [3] should be Default (in our code called Pointer),
since [3] has user-select-none.

* Source/WebCore/html/shadow/imageOverlay.css:
(div#image-overlay):
Adding cursor:auto to div#image-overlay so we can hint text selection with IBeam cursor.

Canonical link: https://commits.webkit.org/253823@main
